lib/ansible/utils/collection_loader.py

```diff
@@ -0,0 +1,305 @@
+
+                sys.modules[fullname] = mod
+
+                return mod
+            elif pkg_type == 'flatmap':
+                raise NotImplementedError()
+            elif pkg_type == 'pkg_only':
+                newmod = ModuleType(fullname)
+                newmod.__package__ = fullname
+                newmod.__file__ = '<ansible_synthetic_collection_package>'
+                newmod.__loader__ = self
+                newmod.__path__ = []
+
+                sys.modules[fullname] = newmod
+
+                return newmod
+
+        if not parent_pkg:  # top-level package, look for NS subpackages on all collection paths
+            package_paths = [self._extend_path_with_ns(p, fullname) for p in self._collection_paths]
+        else:  # subpackage; search in all subpaths (we'll limit later inside a collection)
+            package_paths = [self._extend_path_with_ns(p, fullname) for p in parent_pkg.__path__]
+
+        for candidate_child_path in package_paths:
+            source = None
+            is_package = True
+            location = None
+            # check for implicit sub-package first
+            if os.path.isdir(candidate_child_path):
+                # Py3.x implicit namespace packages don't have a file location, so they don't support get_data
+                # (which assumes the parent dir or that the loader has an internal mapping); so we have to provide
+                # a bogus leaf file on the __file__ attribute for pkgutil.get_data to strip off
+                location = os.path.join(candidate_child_path, '__synthetic__')
+            else:
+                for source_path in [os.path.join(candidate_child_path, '__init__.py'),
+                                    candidate_child_path + '.py']:
+                    if os.path.isfile(source_path):
+                        with open(source_path, 'rb') as fd:
+                            source = fd.read()
+                        location = source_path
+                        is_package = source_path.endswith('__init__.py')
+                        break
+                if not location:
+                    continue
+
+            newmod = ModuleType(fullname)
+            newmod.__package__ = fullname
+            newmod.__file__ = location
+            newmod.__loader__ = self
+
+            if is_package:
+                if sub_collection:  # we never want to search multiple instances of the same collection; use first found
+                    newmod.__path__ = [candidate_child_path]
+                else:
+                    newmod.__path__ = package_paths
+
+            if source:
+                # FIXME: decide cases where we don't actually want to exec the code?
+                exec (source, newmod.__dict__)
+
+            sys.modules[fullname] = newmod
+
+            return newmod
+
+        # FIXME: need to handle the "no dirs present" case for at least the root and synthetic internal collections like ansible.core
+
```

Comment:
Wondering if it would be useful to get the resolved info (the map of what ansible_collections.* import ended up resolving to) into a serializable form?

If there is support for serializing that, and support for creating an AnsibleCollectionLoader from that data structure, that could be used to implement persisting that info. 

If that info can be persisted, it could be used for: 
 - caching between invocations
 - saving the results for use in debugging/troubleshooting/support/auditing
 - reproducing ansible executions (either for repeatability in use, or for reproducing issues)
 - Allowing a very specific manifest of collections/plugins to be loaded, potentially tweaked edited by hands/tooling
   (sort of a collection 'virtualenv' or lockfile)
 - Potentially useful for figuring out the minimal deps a particular playbook invocation needs, so a min content root
   can be constructed. For ex, building a custom 'meta' collection that has exact deps on all the collections used.